website/docs/cdktf/concepts/constructs.mdx

```diff
@@ -8,25 +8,315 @@ description: >-
+
+```ts
+import { Construct } from "constructs";
+
+class Cluster extends Construct {
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+  }
+
+  addService(webservice: Webservice) {
+    // e.g. register a service in a k8s cluster
+  }
+}
+
+class Webservice extends Construct {
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+  }
+}
+
+class MyStack extends TerraformStack {
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+
+    const cluster = new Cluster(this, "cluster");
+    const webservice = new Webservice(this, "webservice");
+
+    cluster.addService(webservice);
+  }
+}
+```
+
+### You can codify architectural patterns
+
+If you have complex architectural patterns you can codify them and have an abstraction where you can create a flexible architecture with less work. In this DataPipeline example you can see that each of the steps in the pipeline are bringing their own logic on how to connect to the next step and how to write their output into an S3Bucket. This means as long as a set of resources can fulfil this contract one could build a pipeline using e.g. LambdaFunctions, Ec2 Instances, and CloudwatchEvents if one implemented the `PipelineStep` class for these types.
+
+```ts
+abstract class PipelineStep extends Construct {
+  abstract connect(next: PipelineStep);
+  abstract finalize(sink: S3Bucket);
+}
+
+class DataPipeline extends Construct {
+  _nextStep: PipelineStep;
+
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+    // this is always the first step
+    this._nextStep = new PipelineWebhookStep(this);
+  }
+
+  public addStep(step: PipelineStep) {
+    this._nextStep.connect(step);
+    this.addAccessPermissions(this._nextStep, step);
+    this._nextStep = step;
+  }
+
+  public finalize() {
+    this._nextStep.finalize();
+  }
+}
+
+class MyStack extends TerraformStack {
+  constructor(scope: Construct, name: string) {
+    super(scope, name);
+    const pipeline = new DataPipeline(this, "my-pipeline");
+    pipeline.addStep(new TransformInputPipelineStep(this));
+    pipeline.addStep(new AggregateValuesPipelineStep(this));
+    pipeline.addStep(new PipelineStoreResultStep(this));
+    pipeline.finalize();
+  }
+}
+```
+
+### You can codify security constraints earlier in the process
+
+You can improve the feedback cycles about security concerns for developers from tools later in the workflow like code review or CI solutions like sentinel to the point where the cofiguration is authored. This can be done through type hints (disallowing options that are not save on the type system level) or through validation. Validations allow dynamic verifications and are shown to users as hints during `cdktf synth`, if they are on the error level they also fail the synth.
```

Comment:
```suggestion
You can improve the feedback cycles about security concerns for developers from tools later in the workflow like code review or CI solutions like Sentinel to the point where the configuration is authored. This can be done through type hints (disallowing options that are not safe on the type system level) or through validation. Validations allow dynamic verification and are shown to users as hints during `cdktf synth`, if they are on the error level they also fail the synth.
```